core/src/main/java/com/linecorp/armeria/client/EventLoopScheduler.java

```diff
@@ -0,0 +1,304 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Streams;
+
+import com.linecorp.armeria.common.util.Releasable;
+
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+
+final class EventLoopScheduler {
+
+    private final EventLoop[] eventLoops;
+    private final Map<String, State> map = new ConcurrentHashMap<>();
+    private int counter;
+
+    EventLoopScheduler(EventLoopGroup eventLoopGroup) {
+        eventLoops = Streams.stream(eventLoopGroup)
+                            .map(e -> (EventLoop) e)
+                            .toArray(EventLoop[]::new);
+    }
+
+    Entry acquire(Endpoint endpoint) {
+        requireNonNull(endpoint, "endpoint");
+        final State state = state(endpoint);
+        final Entry acquired = state.acquire();
+        removeInactiveStates();
+        return acquired;
+    }
+
+    @VisibleForTesting
+    Entry[] entries(Endpoint endpoint) {
+        return state(endpoint).entries();
+    }
+
+    private State state(Endpoint endpoint) {
+        final String authority = endpoint.authority();
+        return map.computeIfAbsent(authority, e -> new State(eventLoops));
+    }
+
+    /**
+     * Cleans up empty entries with no activity for last 1 minute. For reduced overhead, we perform this on
+     * every 1000th acquisition approximately.
+     */
+    private void removeInactiveStates() {
+        if (++counter % 1000 != 0) {
+            return;
+        }
+
+        final long currentTimeNanos = System.nanoTime();
+        for (Iterator<State> i = map.values().iterator(); i.hasNext();) {
+            final State state = i.next();
+            final boolean remove;
+
+            synchronized (state) {
+                remove = state.numAllChannels == 0 &&
+                         currentTimeNanos - state.lastActivityTimeMillis > 60_000_000_000L;
+            }
+
+            if (remove) {
+                i.remove();
+            }
+        }
+    }
+
+    private static final class State {
+        /**
+         * A binary heap of Entry. Ordered by:
+         * <ul>
+         *   <li>{@link Entry#numChannels()} (lower is better)</li>
+         *   <li>{@link Entry#id()} (lower is better)</li>
+         * </ul>
+         */
+        private final Entry[] entries;
+        private int numAllChannels;
+        private long lastActivityTimeMillis = System.nanoTime(); // Updated only when numAllChannels == 0
+
+        State(EventLoop[] eventLoops) {
+            eventLoops = shuffledCopy(eventLoops);
```

Comment:
I'm curious if we should add entries lazily. For example, with 200 event loops and 6000 endpoints, we will end up with 1,200,000 (200 * 6000) Entry instances.